src/coreclr/gc/wasm/gcenv.wasm.cpp

```diff
@@ -0,0 +1,181 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "common.h"
+#include "gcenv.h"
+
+// Emscripten does not provide a complete implementation of mmap and munmap: munmap cannot unmap partial allocations
+// Emscripten does provide an implementation of posix_memalign which is used here.
+
+struct _reserved
+{
+    void * address;
+    size_t size;
+    int used;
+};
+
+typedef struct _reserved reserved;
+
+static int nextPos = 0;
+static reserved blocks[1000]; // better to allocate and grow this as needed (until malloc fails)?
```

Comment:
This is really only used for argument validation. The GC should not be ever calling these APIs with invalid arguments.

I think you can omit this structure, and instead directly call `posix_memalign` / `free` / `memset` to implement `reserve` / `release` / `decommit`.

Reply:
Thanks, works with without that.